compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -4,1440 +4,376 @@
+struct LayoutInfo {
+  /// Given a value, propagate it's layout information forward through it's
+  /// users.
+  void propagateLayoutForward(Value val);
+  /// Given a value, propagate it's layout information backward through it's
+  /// defining operation.
+  void propagateLayoutBackward(Value val);
 
-  /// The result of a resolution.
-  /// Change: The layout was changed.
-  /// Conflict: The layout was not changed because there was a conflict.
-  /// NoChange: The layout was not changed because it was already the same.
-  enum ResolutionResult {
-    Change,
-    Conflict,
-    NoChange,
-  };
+  void setLayoutIfUnset(Value val, VectorLayoutInterface layout) {
+    if (!isa<ShapedType>(val.getType())) {
+      // Don't set layouts on non-shaped types. This would anyway be an empty
+      // layout.
+      return;
+    }
+    if (hasLayout(val)) {
+      return;
+    }
+    layouts[val] = layout;
+    forward.push(val);
+    backward.push(val);
+  }
+  void setLayoutOrClone(OpOperand *val, VectorLayoutInterface layout) {
+    if (!isa<ShapedType>(val->get().getType())) {
+      // Don't set layouts on non-shaped types. This would anyway be an empty
+      // layout.
+      return;
+    }
+    // Always clone constant like ops and set the layout on them.
+    OpBuilder b(val->getOwner());
+    Operation *defOp = val->get().getDefiningOp();
+    if (llvm::isa_and_nonnull<arith::ConstantOp, vector::StepOp,
+                              vector::CreateMaskOp, vector::ConstantMaskOp>(
+            defOp)) {
+      b.setInsertionPoint(defOp);
+      Operation *cloned = b.clone(*defOp);
+      val->set(cloned->getResult(0));
+      layouts[cloned->getResult(0)] = layout;
+      return;
+    }
 
-  /// Attempt to resolve the current lattice with the given lattice. Returns if
-  /// the current layout was not changed, changed or if there was a layout
-  /// conflict.
-  ResolutionResult doResolution(const VectorLayoutInterface &rhs);
+    if (!hasLayout(val->get())) {
+      layouts[val->get()] = layout;
+      forward.push(val->get());
+      backward.push(val->get());
+      return;
+    }
 
-  /// Set the layout for this lattice element to the given layout. This function
-  /// should only be used when you know there will be no layout conflicts.
-  /// Otherwise, the resolve-like functions should be used.
-  void setInnerLayout(const VectorLayoutInterface &layout) {
-    assert(layout &&
-           layout.isValidLayout(getValue().getType(), getValue().getLoc())
-               .succeeded());
-    vectorLayout = layout;
+    // Otherwise, create a to_layout op to change the layout.
+    Value v = val->get();
+    Value layourtedV = ToLayoutOp::create(b, v.getLoc(), v, layout);
+    val->set(layourtedV);
+    layouts[layourtedV] = layout;
+    return;
   }
+  VectorLayoutInterface getLayout(Value val) { return layouts.lookup(val); }
+  bool hasLayout(Value val) { return layouts.contains(val); }
```

Comment:
```suggestion
  bool hasLayout(Value val) const { return layouts.contains(val); }
```

Also, I guess you could make functions `setLayoutOrClone` etc private, although I suppose this struct is more just an internal detail needed to share the graph state (forward/backward) across functions.

Reply:
It's an internal class anyay, there is really no need for these things to be private

Reply:
Point taken, but in general being internal is definitely not a reason to not use `private`. Using `private` makes it immediately clear to the user how objects are used/called.